### PR TITLE
[mplex] Small enhancements.

### PR DESCRIPTION
**Pull request by romanb**
This PR builds on https://github.com/libp2p/rust-libp2p/pull/1784 (proper diff [here](https://github.com/romanb/rust-libp2p/compare/mplex-split-buffer...romanb:mplex-polish?expand=1)), making a few further small improvements:

  * Replace use of `fnv` with `nohash-hasher` to avoid unnecessary hashing of substream IDs.
  * Improve logging through randomly generated connection IDs in the same way as in Yamux.
  * More granular execution of pending flushes for newly opened outbound streams before reading. Otherwise there is a theoretical possibility of both ends being unnecessarily blocked on reading due to attempting to flush and both sides cannot flush for some reason. It is only necessary to flush before a read if the particular substream for which the read happens has just been opened, but not for other existing substreams.

### CODE REVIEW

File: muxers/mplex/src/io.rs [hunk 1]
```diff
@@ -22,22 +22,42 @@ use bytes::Bytes;
...
+use std::{cmp, fmt, io, mem, sync::Arc, task::{Context, Poll, Waker}};
 
 pub use std::io::{Result, Error, ErrorKind};
 
+/// A connection identifier.
+///
+/// Randomly generated and mainly intended to improve log output
+/// by scoping substream IDs to a connection.
+#[derive(Clone, Copy)]
+struct Id(u32);
```
Comment (mxinden) [inline muxers/mplex/src/io.rs:41]: I find it a bit confusing that `Id` is an identifier for a (multiplexed) stream whereas `LocalStreamId` is an identifier for a substream. Would it not be more descriptive if we had `StreamId` and `SubstreamId` ( / `LocalSubstreamId`)?

Comment (mxinden) [inline muxers/mplex/src/io.rs:41]: If I am not mistaken with the birthday paradox in mind on a server with ~100_000 connections the chance for a collision is > 50%. While these identifiers are only used for logging (for now) is the expected performance win with a u32 instead of a u64 worth the risk of a collision?

Comment (romanb) [reply muxers/mplex/src/io.rs:41]: It's not a matter of performance but of compact representation in the logs. Compare:
```rust
[2020-10-14T11:01:44Z DEBUG libp2p_mplex::io] f00c8822: New inbound substream: (18/receiver) (total 19)
[2020-10-14T11:01:44Z TRACE libp2p_mplex::io] f00c8822: Received Open { stream_id: RemoteStreamId { num: 19, role: Dialer } }
```
with
```rust
[2020-10-14T11:02:40Z TRACE libp2p_mplex::io] 19350c447f51a00a: Received Data { stream_id: RemoteStreamId { num: 0, role: Dialer }, data: b"Hello world" }
[2020-10-14T11:02:40Z TRACE libp2p_mplex::io] 19350c447f51a00a: Buffering b"Hello world" for stream (0/receiver) (total: 139)
```
I don't mind using a `u64` instead if you don't think the long representation in the logs is too distracting. We could also use a `u64` but case to `u32` for logging, but that may just add confusion.

Comment (romanb) [reply muxers/mplex/src/io.rs:41]: I'm not really happy with either naming and struggled with it. "Stream" is certainly an appropriate and general term that can be used in all these cases and what kind of stream it is depends on the context. Making that context completely unambiguous in the name is tricky. The relative prefix "sub" only seems to help because in the context of libp2p we are mainly concerned with connections with one level of multiplexing. The API of this module currently uses the terminology `Multiplexed::poll_open_stream` etc. because it is clear from the context that these are "sub" streams of a multiplexed stream. Admittedly in the documentation and logs I've added the "sub" prefix for consistency with the public `StreamMuxer` API and terminology. The `Multiplexed` API on the other hand is only internal. All that said, would you go along for now with `Id` -> `ConnectionId`?

Comment (romanb) [reply muxers/mplex/src/io.rs]: What do you think of https://github.com/libp2p/rust-libp2p/pull/1785/commits/f6bb3544128f961a23261a6fb71a56cd4504c901?

Comment (romanb) [reply muxers/mplex/src/io.rs]: What do you think of https://github.com/libp2p/rust-libp2p/pull/1785/commits/f6bb3544128f961a23261a6fb71a56cd4504c901?

Comment (mxinden) [reply muxers/mplex/src/io.rs]: > The Multiplexed API on the other hand is only internal. All that said, would you go along for now with Id -> ConnectionId?

Yes. I am in favor of `ConnectionId` as it mirrors the [`core` `ConnectionId`](https://github.com/libp2p/rust-libp2p/blob/master/core/src/connection/manager.rs#L77).

Comment (mxinden) [reply muxers/mplex/src/io.rs]: I am not in favor of introducing a global variable. While still only used for logging, I would find it confusing that connection ids are monotonic e.g. when executing tests in serial, and possibly not monotonic (per test) when run concurrently.

Even though it is a bit verbose, I would prefer the u64, simply as there are no edge-cases. But I am fine with the u32 as well (maybe with a comment pointing out the probability of collisions).

Comment (romanb) [reply muxers/mplex/src/io.rs]: > I am not in favor of introducing a global variable. While still only used for logging, I would find it confusing that connection ids are monotonic e.g. when executing tests in serial, and possibly not monotonic (per test) when run concurrently.

I don't quite understand your concerns with monotonic connection IDs or why you are referring to a module-scoped atomic counter as a global variable, especially in contrast to the use of a global random number generator, but I'm OK with the random `u64` for now.

Comment (romanb) [reply muxers/mplex/src/io.rs]: See https://github.com/libp2p/rust-libp2p/pull/1785/commits/b98234b3dc6f1ae80cc071006dfaf9cca81e119c.

**Review by mxinden — APPROVED** [review]
Looks good to me.
